hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/HoodieWriteHelper.java

```diff
@@ -78,7 +79,8 @@ protected HoodieData<HoodieRecord<T>> doDeduplicateRecords(
         throw new HoodieException(String.format("Error to merge two records, %s, %s", rec1, rec2), e);
       }
       HoodieKey reducedKey = rec1.getData().equals(reducedRecord.getData()) ? rec1.getKey() : rec2.getKey();
-      return reducedRecord.newInstance(reducedKey);
+      HoodieOperation operation = rec1.getData().equals(reducedRecord.getData()) ? rec1.getOperation() : rec2.getOperation();
+      return reducedRecord.newInstance(reducedKey, operation);
```

Comment:
Does the code make sense to you?

```java
      boolean choosePrev = rec1.getData().equals(reducedRecord.getData());
      return choosePrev
          ? reducedRecord.newInstance(rec1.getKey(), rec1.getOperation())
          : reducedRecord.newInstance(rec2.getKey(), rec2.getOperation());
```

Actually, the current contract of payload merging is either one of the payload would be returned(no copying, no composition), so we can use objects address equals `==` which is far more efficient. Something like the following code:

```java
      boolean choosePrev = reducedRecord == rec1;
      return choosePrev ? rec1 : rec2;
```

Reply:
not sure this statement "the current contract of payload merging is either one of the payload would be returned(no copying, no composition)" is true? 
for partial update payload scenarios, new payload constructed will include some values from old record and some from new incoming.

Reply:
> for partial update payload scenarios, new payload constructed will include some values from old record and some from new incoming

Yeah, so new object is returns and the decision for `==` always returns false, so this patch is kind of conservative and still keeps the `equals` comparison for the payload, but for Flink, we just use the object `==` directly.